src/components/web-form-renderer.vue

```diff
@@ -0,0 +1,336 @@
+  submissionModal.state = true;
+  submissionModal.type = `${type}Modal`;
+};
+
+const hideSubmissionModal = () => {
+  submissionModal.hide();
+  primaryInstance.instanceId = null;
+};
+
+/**
+ * Web Form expects host application to provide a function that it can use to
+ * fetch attachments. Signature of the function is (url) => Response; where
+ * Response is subset of web standard  {@link Response}.
+ */
+const getAttachment = (url) => {
+  let requestUrl = apiPaths.formAttachment(
+    form.projectId,
+    form.xmlFormId,
+    form.draftToken !== null,
+    url.pathname.substring(1)
+  );
+  if (form.draftToken) {
+    requestUrl = requestUrl.replace(/^\/v1\//, `/v1/test/${form.draftToken}/`);
+  }
+  if (route.query.st) {
+    requestUrl += `?st=${route.query.st}`;
+  }
+  return request({
+    url: requestUrl,
+    alert: false
+  }).then(axiosResponse => {
+    const { data, status, statusText, headers } = axiosResponse;
+
+    const fetchHeaders = new Headers();
+    for (const [key, value] of Object.entries(headers)) {
+      if (key === 'content-type') {
+        // because web-forms doens't want space between media type and charset
+        // https://github.com/getodk/web-forms/issues/269
+        fetchHeaders.append(key, value.replace('; charset', ';charset'));
+      } else {
+        fetchHeaders.append(key, value);
+      }
+    }
+
+    let body;
+    if (typeof (data) === 'string') {
+      body = data;
+    } else if (headers['content-type'].includes('application/json') ||
+              headers['content-type'].includes('application/geo+json')) {
+      body = JSON.stringify(data);
+    } else {
+      // eslint-disable-next-line no-console
+      console.error('response data is not a known text format');
+    }
+
+    return new Response(body, {
+      status,
+      statusText,
+      headers: fetchHeaders,
+    });
+  });
+};
+
+const postPrimaryInstance = () => {
+  let url = apiPaths.submissions(form.projectId, form.xmlFormId, form.draftToken, '');
+  if (form.draftToken) {
+    url = url.replace(/^\/v1\//, `/v1/test/${form.draftToken}/`);
+  }
+  if (route.query.st) {
+    url += `?st=${route.query.st}`;
+  }
+  return request({
+    method: 'POST',
+    url,
+    data: primaryInstance.file,
+    headers: {
+      'content-type': 'text/xml'
+    },
+    fulfillProblem: () => true
+  })
+    .then(({ data }) => {
+      if (isProblem(data)) {
+        primaryInstance.uploadSuccess = false;
+        submissionModal.errorMessage = data.message;
+        showModal('error');
+        return false;
+      }
+      Object.assign(primaryInstance, { instanceId: data.instanceId, uploadSuccess: true });
+      return true;
+    })
+    .catch(noop);
+};
+
+const uploadAttachment = async (attachmentIndex) => {
+  const attachmentDescriptor = attachmentUploads[attachmentIndex];
+  let url = apiPaths.submissionAttachment(form.projectId, form.xmlFormId, false, primaryInstance.instanceId, attachmentDescriptor.file.name);
```

Comment:
The third argument to `apiPaths.submissionAttachment()` doesn't look right to me. Isn't it possible for this to be a form draft?